kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver_with_constraints.h

```diff
@@ -0,0 +1,1142 @@
+        LocalSystemVectorType constant_vector = LocalSystemVectorType(0);
+        EquationIdVectorType slave_equation_ids = EquationIdVectorType(0);
+        EquationIdVectorType master_equation_ids = EquationIdVectorType(0);
+
+#pragma omp parallel for schedule(guided, 512) firstprivate(number_of_constraints, constraints_begin, relation_matrix, constant_vector, slave_equation_ids, master_equation_ids)
+        for (int i_constraints = 0; i_constraints < number_of_constraints; i_constraints++)
+        {
+            ModelPart::MasterSlaveConstraintContainerType::iterator it = constraints_begin + i_constraints;
+
+            //detect if the element is active or not. If the user did not make any choice the element
+            //is active by default
+            bool constraint_is_active = true;
+            if ((it)->IsDefined(ACTIVE))
+                constraint_is_active = (it)->Is(ACTIVE);
+
+            if (constraint_is_active)
+            {
+                //get the equation Ids of the constraint
+                (*it).EquationIdVector(slave_equation_ids, master_equation_ids, r_current_process_info);
+                //calculate constraint's T and b matrices
+                (*it).CalculateLocalSystem(relation_matrix, constant_vector, r_current_process_info);
+
+                //assemble the Constraint contribution
+                AssembleSlaves(slave_equation_ids, master_equation_ids, relation_matrix);
+            }
+        }
+        const double stop_formulate = OpenMPUtils::GetCurrentTime();
+        KRATOS_INFO_IF("ResidualBasedBlockBuilderAndSolverWithConstraints", (this->GetEchoLevel() >= 1 && rModelPart.GetCommunicator().MyPID() == 0)) << "Formulate global constraints time: " << stop_formulate - start_formulate << std::endl;
+
+        KRATOS_CATCH("ResidualBasedBlockBuilderAndSolverWithConstraints::FormulateGlobalMasterSlaveRelations failed ..");
+    }
+
+
+    /**
+     * @brief   this method assembles the given list of slave equation ids and corresponding master equatio ids
+     * @param   rSlaveEquationIdVector list of slave equation ids to be assembled.
+     * @param   rMasterEquationIdVector list of corresponding master equation ids.
+     * @param   rRelationMatrix the matrix relating the rSlaveEquationIdVector with rMasterEquationIdVector
+     */
+    void AssembleSlaves(EquationIdVectorType& rSlaveEquationIdVector, EquationIdVectorType& rMasterEquationIdVector, LocalSystemMatrixType& rRelationMatrix)
+    {
+        KRATOS_TRY
+        int slave_count = 0;
+        for (auto &slave_equation_id : rSlaveEquationIdVector)
+        {
+            int master_count = 0;
+            auto global_constraint = mGlobalMasterSlaveRelations.find(slave_equation_id);
+            if (global_constraint == mGlobalMasterSlaveRelations.end())
+            {
+                mGlobalMasterSlaveRelations.insert(mGlobalMasterSlaveRelations.begin(), Kratos::make_shared<AuxilaryGlobalMasterSlaveRelationType>(slave_equation_id));
```

Comment:
why using shared_ptr now? i think each will be unique right? if so use unique_ptr.

Reply:
I am using pointer_vector_set to store the the unique global m-s relations. I am not sure if pointer_vector_set can work with unique_pointers.

Reply:
PointerVectorSet does not currently work with unique PTR as of now (see bug #2198, it is smthg similar to it ... @pooyan-dadvand do you agree or am i wrong on this one?)

still i would go for a std::unordered_map(constraint_id, unique_ptr< AuxiliaryGlobal.......> )
while we get the issue with PointerVector smoothed